# MovieLens with SAP HANA APL Recommendation


## **Initialize the connection**

In [1]:
import sqlalchemy, os
from sqlalchemy import create_engine

%reload_ext sql
%config SqlMagic.displaylimit = 5
%config SqlMagic.feedback = False
%config SqlMagic.autopandas = True

hxe_connection = 'hana://ML_USER:Welcome18@hxehost:39015';

%sql $hxe_connection

import pandas as pd
pd.options.display.max_rows = 100

from IPython.core.display import HTML
HTML("""
<script>
code_show=true; 
function code_toggle() {
    if (code_show){
        $('div.cell.code_cell.rendered.selected div.input').hide();
    } else {
        $('div.cell.code_cell.rendered.selected div.input').show();
    }
    code_show = !code_show
} 

$( document ).ready(code_toggle);
</script>

To show/hide this cell's raw code input, click <a href="javascript:code_toggle()">here</a>.""")

## **Cleanup Model Tables and APL objects**

In [2]:
%%sql
-- --------------------------------------------------------------------------
-- cleanup sapl objects
-- --------------------------------------------------------------------------
call sap_pa_apl."sap.pa.apl.base::CLEANUP"(1,?);
-- --------------------------------------------------------------------------
-- drop function in/out tables, helper tables and views
-- --------------------------------------------------------------------------
drop table apl_movielens_function_header;
drop table apl_movielens_parameters;
drop table apl_movielens_variable_desc;
drop table apl_movielens_operation_log;
drop table apl_movielens_summary;
drop table apl_movielens_indicators;
drop table apl_movielens_operation_result;
drop table apl_movielens_model;
drop table apl_movielens_model_node_users;
drop table apl_movielens_model_node_items;
drop table apl_movielens_model_links;

 * hana://ML_USER@hxehost:39015


Empty DataFrame
Columns: []
Index: []

## **Create Model Table **

In [3]:
%%sql
-- --------------------------------------------------------------------------
-- create generic tables using pre-built table types
-- --------------------------------------------------------------------------
create column table apl_movielens_function_header   like sap_pa_apl."sap.pa.apl.base::BASE.T.FUNCTION_HEADER";
create column table apl_movielens_parameters        like sap_pa_apl."sap.pa.apl.base::BASE.T.OPERATION_CONFIG_EXTENDED";
create column table apl_movielens_variable_desc     like sap_pa_apl."sap.pa.apl.base::BASE.T.VARIABLE_DESC_OID";
create column table apl_movielens_operation_log     like sap_pa_apl."sap.pa.apl.base::BASE.T.OPERATION_LOG";
create column table apl_movielens_summary           like sap_pa_apl."sap.pa.apl.base::BASE.T.SUMMARY";
create column table apl_movielens_indicators        like sap_pa_apl."sap.pa.apl.base::BASE.T.INDICATORS";
create column table apl_movielens_operation_result  like sap_pa_apl."sap.pa.apl.base::BASE.T.RESULT";
create column table apl_movielens_model             like sap_pa_apl."sap.pa.apl.base::BASE.T.MODEL_NATIVE";
-- --------------------------------------------------------------------------
-- create model tables
-- --------------------------------------------------------------------------
create column table apl_movielens_model_node_users (
    node integer    -- must be of the same sql type as the user column (userid here)
);
create column table apl_movielens_model_node_items (
    node integer    -- must be of the same sql type as the item column (movieid here)
);
create column table apl_movielens_model_links (
    graph_name        nvarchar(255),
    weight            double,
    kxnodefirst       integer,    -- must be of the same sql type as the user column (userid here)
    kxnodesecond      integer,    -- must be of the same sql type as the item column (movieid here)
    kxnodesecond_2    integer     -- must be of the same sql type as the item column (movieid here)
);

 * hana://ML_USER@hxehost:39015


Empty DataFrame
Columns: []
Index: []

## **Run the Recommendation algorithm**

In [4]:
%%sql
-- --------------------------------------------------------------------------
-- configuration
-- --------------------------------------------------------------------------
truncate table apl_movielens_function_header;
insert into apl_movielens_function_header values ('Oid', '#1');
insert into apl_movielens_function_header values ('LogLevel', '8');

truncate table apl_movielens_parameters;
insert into apl_movielens_parameters values ('APL/ModelType'              , 'recommendation'  , null);
insert into apl_movielens_parameters values ('APL/User'                   , 'USERID'          , null); -- mandatory
insert into apl_movielens_parameters values ('APL/Item'                   , 'MOVIEID'         , null); -- mandatory
insert into apl_movielens_parameters values ('APL/RuleWeight'             , 'Support'         , null); -- default is Independence Probability
insert into apl_movielens_parameters values ('APL/BestSeller'             , '50000'           , null); -- default is 50000
insert into apl_movielens_parameters values ('APL/MaxTopNodes'            , '100000'          , null); -- default is 100000
insert into apl_movielens_parameters values ('APL/MinimumConfidence'      , '0.05'            , null); -- default is 0.05
insert into apl_movielens_parameters values ('APL/MinimumPredictivePower' , '0.0'             , null); -- default is 0.0
insert into apl_movielens_parameters values ('APL/MinimumSupport'         , '2'               , null); -- default is 2
insert into apl_movielens_parameters values ('APL/Top'                    , '5'               , null); -- default is max
insert into apl_movielens_parameters values ('APL/IncludeBestSellers'     , 'false'           , null); -- default is false

select * from apl_movielens_parameters;

 * hana://ML_USER@hxehost:39015


key           value context
0                APL/ModelType  recommendation    None
1                     APL/User          USERID    None
2                     APL/Item         MOVIEID    None
3               APL/RuleWeight         Support    None
4               APL/BestSeller           50000    None
5              APL/MaxTopNodes          100000    None
6        APL/MinimumConfidence            0.05    None
7   APL/MinimumPredictivePower             0.0    None
8           APL/MinimumSupport               2    None
9                      APL/Top               5    None
10      APL/IncludeBestSellers           false    None

In [5]:
%%sql
-- --------------------------------------------------------------------------
-- Clean result tables
-- --------------------------------------------------------------------------
truncate table apl_movielens_model;
truncate table apl_movielens_model_node_users;
truncate table apl_movielens_model_node_items;
truncate table apl_movielens_model_links;
-- --------------------------------------------------------------------------
-- execute the apl function to train the model
-- --------------------------------------------------------------------------
call sap_pa_apl."sap.pa.apl.base::CREATE_RECO_MODEL_AND_TRAIN" (
    apl_movielens_function_header
  , apl_movielens_parameters
  , apl_movielens_variable_desc
  , current_schema, 'MOVIELENS_RATINGS'
  , apl_movielens_model
  , current_schema, 'APL_MOVIELENS_MODEL_NODE_USERS'
  , current_schema, 'APL_MOVIELENS_MODEL_NODE_ITEMS'
  , current_schema, 'APL_MOVIELENS_MODEL_LINKS'
  , apl_movielens_operation_log
  , apl_movielens_summary
  , apl_movielens_indicators
  , apl_movielens_operation_result
) with overview;

 * hana://ML_USER@hxehost:39015


variable                                       table
0      INDICATOR        "ML_USER"."APL_MOVIELENS_INDICATORS"
1          MODEL             "ML_USER"."APL_MOVIELENS_MODEL"
2  OPERATION_LOG     "ML_USER"."APL_MOVIELENS_OPERATION_LOG"
3         RESULT  "ML_USER"."APL_MOVIELENS_OPERATION_RESULT"
4        SUMMARY           "ML_USER"."APL_MOVIELENS_SUMMARY"

## **Check the output**

In [6]:
%sql select * from apl_movielens_operation_log;

 * hana://ML_USER@hxehost:39015


oid               timestamp  level                      origin  \
0   #1 2018-07-20 14:03:14.420      3                    APLModel   
1   #1 2018-07-20 14:03:14.420      3                    APLModel   
2   #1 2018-07-20 14:03:14.420      3                    APLModel   
3   #1 2018-07-20 14:03:14.420      3                    APLModel   
4   #1 2018-07-20 14:03:14.420      3         Kxen.Recommendation   
5   #1 2018-07-20 14:03:14.420      3         Kxen.Recommendation   
6   #1 2018-07-20 14:03:14.836      3  cWithContext[7f90f1c50898]   
7   #1 2018-07-20 14:03:14.836      3         Kxen.Recommendation   
8   #1 2018-07-20 14:03:14.836      3  cWithContext[7f911873f5d8]   
9   #1 2018-07-20 14:03:14.836      3  cWithContext[7f911873f5d8]   
10  #1 2018-07-20 14:03:49.928      3  cWithContext[7f911873f5d8]   
11  #1 2018-07-20 14:03:49.977      3  cWithContext[7f911873f5d8]   
12  #1 2018-07-20 14:03:49.977      3  cWithContext[7f911873f5d8]   
13  #1 2018-07-20 14:03:50.000      3  cWithContext[7f911873f5d8]   
14  #1 2018-07-20 14:03:50.000      3  cWithContext[7f911873f5d8]   
15  #1 2018-07-20 14:03:50.011      3  cWithContext[7f911873f5d8]   
16  #1 2018-07-20 14:03:50.011      3  cWithContext[7f911873f5d8]   
17  #1 2018-07-20 14:03:50.011      3  cWithContext[7f911873f5d8]   
18  #1 2018-07-20 14:03:50.017      3  cWithContext[7f911873f5d8]   
19  #1 2018-07-20 14:03:50.017      3  cWithContext[7f911873f5d8]   
20  #1 2018-07-20 14:03:50.018      4  cWithContext[7f911873f5d8]   
21  #1 2018-07-20 14:03:50.018      4  cWithContext[7f911873f5d8]   
22  #1 2018-07-20 14:03:50.018      4  cWithContext[7f911873f5d8]   
23  #1 2018-07-20 14:03:50.018      4  cWithContext[7f911873f5d8]   
24  #1 2018-07-20 14:03:50.018      3         Kxen.Recommendation   
25  #1 2018-07-20 14:03:50.021      3                    APLModel   
26  #1 2018-07-20 14:03:50.852      1         Kxen.Recommendation   

                                              message  
0                            Checking internal state.  
1                      Total elapsed time: 0 seconds.  
2                            Checking internal state.  
3                  Beginning of learning for Default.  
4        Learn date of the model: 2018-07-20 14:03:14  
5                                   Loading graphs...  
6                           Links dataset read in 0 s  
7                Graph memory consumption: 3991596 KB  
8      Starting graph projection Transactions -> Item  
9    Starting the graph projection iteration 1 (of 1)  
10  The batch mode was selected for derived graph ...  
11  Derived graph links batch creation done in 0 s...  
12    Graph projection iteration done in 35 second(s)  
13        Derived graphs computing time: 35 second(s)  
14                Compute global statistics on graphs  
15  Process size after graph statistics computatio...  
16         Statistics computation done in 1 second(s)  
17                Compute global statistics on graphs  
18  Process size after graph statistics computatio...  
19         Statistics computation done in 0 second(s)  
20           Total unique nodes count: 0 (671 + 9066)  
21             Graph 'Item': 1173 nodes, 100000 links  
22  Statistics for graph 'Item':\n-Density: 0.1453...  
23  Power law of degree distribution for graph 'It...  
24                          Learning time: 36 seconds  
25                    Total elapsed time: 36 seconds.  
26  The database where the SQL code will be execut...

In [7]:
%sql select * from apl_movielens_summary;

 * hana://ML_USER@hxehost:39015


oid                         key                value
0   #1                  ModelState         TrainedModel
1   #1            ModelDatasetName           APL_EVENTS
2   #1          ModelVariableCount                    2
3   #1  ModelSelectedVariableCount                    2
4   #1              ModelAvailable                 true
5   #1              ModelBuildDate  2018-07-20 14:03:50
6   #1           ModelLearningTime                   36
7   #1             ModelEngineName  Kxen.Recommendation
8   #1                 ModelAuthor               hxeadm
9   #1               RecoItemCount                 1173
10  #1               RecoRuleCount               100000
11  #1       RecoFilteredLineCount                    0
12  #1      TransDictinctUserCount                  671
13  #1      TransDistinctItemCount                 9066
14  #1      TransPairUserItemCount               100004
15  #1      TransFilteredLineCount               100004
16  #1        TransBestSellerCount                    0
17  #1      AplTrainingElapsedTime                35601
18  #1    AplModelWriteElapsedTime                  766
19  #1       AplDebriefElapsedTime                    0
20  #1         AplTotalElapsedTime                36464

In [8]:
%sql select * from apl_movielens_indicators;

 * hana://ML_USER@hxehost:39015


oid variable target               key       value detail
0  #1     None   None  IndicatorDataset  Estimation   None

In [9]:
%sql select * from apl_movielens_operation_result;

 * hana://ML_USER@hxehost:39015


oid   key                                              value
0  #1  code  SELECT * FROM (\nSELECT "T1".$Key,\n\tROW_NUMB...

## **Collaborative filtering - Create the result view**

In [10]:
%sql drop   view apl_movielens_collaborative_filtering;

 * hana://ML_USER@hxehost:39015


Empty DataFrame
Columns: []
Index: []

In [11]:
%%sql
create view apl_movielens_collaborative_filtering as
select
  userid, rank, t1.movieid, score, title, genres, imdbid, tmdbid
from (
  select
      t1.userid
    , row_number() over(partition by t1.userid order by t1.score desc, t1.consequent desc ) as rank
    , t1.consequent as movieid
    , t1.score      as score
  from (
      select
          t1.userid, t1.consequent, max(t1.score)  as score
      from (
        select
            t1.userid
          , t1.consequent
          , t1.support / ( (coalesce (t2_1.count_antecedent,0) + coalesce (t2_2.count_antecedent,0) ) )  as score -- confidence calculation
        from (
            select
                t1.userid
              , t1.antecedent , t1.consequent
              , t1.support
            from (
              select
                  spacein.userid
                , rules.kxnodesecond   as antecedent
                , rules.kxnodesecond_2 as consequent
                , rules.weight         as support
              from movielens_ratings spacein
              left outer join (select * from apl_movielens_model_links where graph_name = 'Transactions') products on (products.kxnodefirst  = spacein.userid)
              left outer join (select * from apl_movielens_model_links where graph_name = 'Item'        ) rules    on (products.kxnodesecond = rules.kxnodesecond)
              left outer join (select * from apl_movielens_model_links where graph_name = 'Transactions') notin    on (rules.kxnodesecond_2  = notin.kxnodesecond) and (notin.kxnodefirst = spacein.userid)
                where rules.kxnodesecond is not null  and notin.kxnodesecond is null
            ) t1
            union all
            select
                t1.userid
              , t1.antecedent , t1.consequent
              , t1.support
            from (
              select
                  spacein.userid
                , rules.kxnodesecond_2 as antecedent
                , rules.kxnodesecond   as consequent
                , rules.weight         as support
              from movielens_ratings spacein
              left outer join (select * from apl_movielens_model_links where graph_name = 'Transactions') products on (products.kxnodefirst  = spacein.userid)
              left outer join (select * from apl_movielens_model_links where graph_name = 'Item'        ) rules    on (products.kxnodesecond = rules.kxnodesecond_2)
              left outer join (select * from apl_movielens_model_links where graph_name = 'Transactions') notin    on (rules.kxnodesecond    = notin.kxnodesecond) and (notin.kxnodefirst = spacein.userid)
              where rules.kxnodesecond_2 is not null and notin.kxnodesecond is null
            ) t1
        ) t1
        left outer join (select kxnodesecond   as antecedent, cast(count(*) as float) as count_antecedent from apl_movielens_model_links where graph_name ='Transactions' group by kxnodesecond  ) t2_1 on (t1.antecedent = t2_1.antecedent)
        left outer join (select kxnodesecond_2 as antecedent, cast(count(*) as float) as count_antecedent from apl_movielens_model_links where graph_name ='Transactions' group by kxnodesecond_2) t2_2 on (t1.antecedent = t2_2.antecedent)
      ) t1 group by t1.userid,  t1.consequent
  ) t1
) t1
left outer join movielens_movies movies on movies.movieid = t1.movieid
left outer join movielens_links  links  on links.movieid  = t1.movieid
where rank <= 5;

 * hana://ML_USER@hxehost:39015


Empty DataFrame
Columns: []
Index: []

## **Content-based filtering - Create the result view**

In [12]:
%sql drop   view apl_movielens_contentbased_filtering;

 * hana://ML_USER@hxehost:39015


Empty DataFrame
Columns: []
Index: []

In [13]:
%%sql
create view apl_movielens_contentbased_filtering as
select
  t1.movieid, rank, similar_movie, score, title, genres, imdbid, tmdbid
from (
  select
      t1.movieid
    , row_number() over(partition by t1.movieid order by t1.score desc, t1.consequent desc ) as rank
    , t1.consequent as similar_movie
    , t1.score
  from (
      select
          t1.movieid
        , t1.consequent
        , max(t1.score) as score
      from (
        select
            t1.movieid
          , t1.consequent
          , t1.support / (coalesce (t2_1.count_antecedent,0) + coalesce (t2_2.count_antecedent,0) ) as score
        from (
            select
                nodes.node as movieid
              , rules.kxnodesecond   as antecedent
              , rules.kxnodesecond_2 as consequent
              , rules.weight as support
            from
              apl_movielens_model_node_items nodes
            left outer join (select * from apl_movielens_model_links where graph_name = 'Item' ) rules    on (nodes.node = rules.kxnodesecond)
              where rules.kxnodesecond_2 is not null
            union all
            select
                nodes.node as movieid
              , rules.kxnodesecond_2 as antecedent
              , rules.kxnodesecond   as consequent
              , rules.weight as support
            from
              apl_movielens_model_node_items nodes
            left outer join (select * from apl_movielens_model_links where graph_name = 'Item' ) rules    on (nodes.node = rules.kxnodesecond_2)
            where rules.kxnodesecond is not null
        ) t1
        left outer join (select kxnodesecond   as antecedent, cast(count(*) as float) as count_antecedent from apl_movielens_model_links where graph_name ='Transactions' group by kxnodesecond  ) t2_1 on (t1.antecedent = t2_1.antecedent)
        left outer join (select kxnodesecond_2 as antecedent, cast(count(*) as float) as count_antecedent from apl_movielens_model_links where graph_name ='Transactions' group by kxnodesecond_2) t2_2 on (t1.antecedent = t2_2.antecedent)
      ) t1 group by t1.movieid, t1.consequent
  ) t1
) t1
left outer join movielens_movies movies on movies.movieid = t1.similar_movie
left outer join movielens_links  links  on links.movieid  = t1.similar_movie
where rank <= 5;

 * hana://ML_USER@hxehost:39015


Empty DataFrame
Columns: []
Index: []

## **Collaborative filtering - Validate the result**
Let's verify how many users will actually get recommendations using the following SQL:

In [14]:
%%sql
select reco_count, count(1) as user_count
from (
  select userid, max(rank) as reco_count
  from apl_movielens_collaborative_filtering
  group by userid
) group by reco_count order by 1 desc;

 * hana://ML_USER@hxehost:39015


reco_count  user_count
0           5         671

Let's verify how many distinct movies will actually get recommended to a user (part of the top 5 scores) using the following SQL:

In [15]:
%%sql
select
    count(1) as movie_count
  , count(1) * 100 / (select count(1) as cnt from movielens_movies) as movie_ratio
from (
  select movieid
  from apl_movielens_collaborative_filtering
  group by movieid
);

 * hana://ML_USER@hxehost:39015


movie_count movie_ratio
0          181    1.983561

Let's verify how many distinct movies will potentially get recommended to a user (not just the top 5 scores) using the following SQL:

In [16]:
%%sql
select
    count(1) as movie_count
  , count(1) * 100 / (select count(1) as cnt from movielens_movies) as movie_ratio
from (
    select movieid
    from (
      select kxnodesecond  as movieid from apl_movielens_model_links where graph_name = 'Item' group by  kxnodesecond
      union all
      select kxnodesecond_2 as movieid from apl_movielens_model_links where graph_name = 'Item' group by  kxnodesecond_2
    ) group by movieid
);

 * hana://ML_USER@hxehost:39015


movie_count movie_ratio
0         1173   12.854794

All users will receive the requested 5 recommendations. However, only about 13% of the movies (1176 out of the 9,125) will get potentially recommended and only about 1.5% of the movies (149 out of the 9,125) are in the top 5 lists.

## **Content-based filtering - Validate the result**
Let's verify how many movies will actually get recommendations using the following SQL:

In [17]:
%%sql
select reco_count, count(1) as movie_count
from (
  select movieid, max(rank) as reco_count
  from apl_movielens_contentbased_filtering
  group by movieid
) group by reco_count;

 * hana://ML_USER@hxehost:39015


reco_count  movie_count
0           5         1050
1           2           34
2           1           42
3           3           23
4           4           24

Let's verify how many distinct similar movies will actually get recommended to a movie (part of the top 5 scores) using the following SQL:

In [18]:
%%sql
select
    count(1) as movie_count
  , count(1) * 100 / (select count(1) as cnt from movielens_movies ) as movie_ratio
from (
  select movieid
  from apl_movielens_contentbased_filtering
  group by movieid
);

 * hana://ML_USER@hxehost:39015


movie_count movie_ratio
0         1173   12.854794

Let's verify how many rating does the movies with no recommendation have using the following SQL:

In [19]:
%%sql
select rating_count, count(1) as movie_count
from (
  select ratings.movieid, count(1) as rating_count
  from movielens_ratings ratings
  left outer join (
    select movieid
    from (
      select movieid
      from (
        select kxnodesecond  as movieid from apl_movielens_model_links where graph_name = 'Item' group by  kxnodesecond
        union all
        select kxnodesecond_2 as movieid from apl_movielens_model_links where graph_name = 'Item' group by  kxnodesecond_2
      ) group by movieid
    )
  ) t1 on (ratings.movieid = t1.movieid)
  where t1.movieid is null
  group by ratings.movieid
) group by rating_count;

 * hana://ML_USER@hxehost:39015


rating_count  movie_count
0              1         3063
1              2         1202
2              3          755
3              4          550
4              5          397
5              6          305
6              7          206
7              8          187
8              9          156
9             10          162
10            11          133
11            12          101
12            13          112
13            14           98
14            15           73
15            16           68
16            17           70
17            18           60
18            19           64
19            20           47
20            21           34
21            22           22
22            23           18
23            24            7
24            25            2
25            27            1

As you can see, the movies with no recommendations have all less than 27 ratings, and this list include the 3063 movies with only one rating and the 1202 with only 2 ratings.